### PR TITLE
Framework: do not enable redux persistence if we switch users

### PR DESCRIPTION
**Pull request by gwwar**
This PR attempts to disable redux persistence if we are switched as another user. ~~The building blocks are here but @jordwest I'm not sure if the timing is correct for `shouldBoot` to return the correct value.~~ Thanks for fixing that @jordwest  

~~@blowery it looks like initial-state tests were accidentally turned off during the test runner conversion. We may want to consider scanning test directories so we don't need to specify every file in a config.~~

## Testing Instructions
* Start Calypso
* Set console debug to 'calypso:state';
* Refresh the page, and make sure Redux is initialized from browser storage.
* Follow switch user instructions here: (ref: p4TIVU-2OW-p2)
* When switched, redux persistence is disabled. You should see a debug message with the following:
```text
'persist-redux is not enabled, building state from scratch'
```

cc @jordwest @dllh @retrofox @aduth

### CODE REVIEW

File: client/test/config.json [hunk 1]
```diff
@@ -113,6 +113,6 @@
 		"users": {
 			"test": [ "actions", "reducer", "selectors" ]
 		},
-		"test": [ "index" ]
+		"test": [ "index", "initial-state" ]
```
Comment (gziolo) [inline client/test/config.json]: Yes, we missed that one ...
I figured out yesterday: https://github.com/Automattic/wp-calypso/commit/4ed280bf9e706a56a00fe0b36258e44dafda7eda, but it isn't merged yet :)

**Comment by jordwest** [comment]
I've just added a commit that should fix any timing issues. Previously the store was cleared when the user switches, leaving the return value of `shouldBoot` indeterminate at certain points in time. This change replaces the `shouldBoot` function with `isSupportUserSession` that's determined at module startup, then kept for the remainder of that session.

File: client/state/test/initial-state.js [hunk 2]
```diff
@@ -10,17 +11,26 @@ import sinon from 'sinon';
...
+		mockery.registerMock( 'lib/user/support-user-interop', { isSupportUserSession: isSupportUserSession } );
+		mockery.enable( { warnOnUnregistered: false } );
 		localforage = require( 'localforage' );
 		const initialState = require( '../initial-state' );
 		createReduxStoreFromPersistedInitialState = initialState.default;
 		MAX_AGE = initialState.MAX_AGE;
 	} );
 
+	after( () => {
+		mockery.deregisterAll();
```
Comment (gziolo) [inline client/state/test/initial-state.js]: It can be simplified now by using `useMockery` helper. See https://github.com/Automattic/wp-calypso/blob/f5bf33b5206c999ff17475900179fb5a846be73e/client/test/use-mockery.js.

File: client/test/config.json (hunk 1 above)
Comment (gziolo) [inline client/test/config.json]: I removed that change from my PR to avoid merge issues.

**Comment by blowery** [comment]
> @blowery it looks like initial-state tests were accidentally turned off during the test runner conversion. We may want to consider scanning test directories so we don't need to specify every file in a config.

That's the plan once we have everything moved over to the single runner. Right now we have to whitelist things while we have both the new single runner and the old set of `Makefile`s.

File: client/state/test/initial-state.js [hunk 3]
```diff
@@ -10,17 +11,26 @@ import sinon from 'sinon';
...
-	let localforage, createReduxStoreFromPersistedInitialState, MAX_AGE;
+	let localforage, createReduxStoreFromPersistedInitialState, MAX_AGE,
+		isSwitchedUser = false,
+		isSupportUserSession = () => isSwitchedUser;
 
 	require( 'lib/react-test-env-setup' ).auto();
 
 	before( () => {
+		mockery.registerMock( 'lib/user/support-user-interop', { isSupportUserSession: isSupportUserSession } );
+		mockery.enable( { warnOnUnregistered: false } );
```
Comment (blowery) [inline client/state/test/initial-state.js]: This needs to spec `useCleanCache: true` since you can no longer guarentee that what you're mocking hasn't been required by someone else in a previous test run with the single runner. Or just use `useMockery` from `test/use-mockery` and it'll be taken care of for you.

File: client/test/config.json (hunk 1 above)
Comment (gwwar) [inline client/test/config.json]: :+1: Thanks :)

File: client/state/test/initial-state.js (hunk 3 above)
Comment (gwwar) [inline client/state/test/initial-state.js]: Hmm, looks like my sinon stubs for config don't work if I use the `use-mockery` helper. I'll investigate later, but I'm going to use the cache clean option for now.

Comment (gwwar) [inline client/state/test/initial-state.js]: updated in d7d8185

File: client/state/test/initial-state.js (hunk 2 above)
Comment (gwwar) [inline client/state/test/initial-state.js]: updated in d7d8185

**Comment by gwwar** [comment]
Heads up, going to rebase to resolve conflicts.

**Comment by gwwar** [comment]
> Scanning instead of using a whitelist is the plan once we have everything moved over to the single runner. 

:+1: gotcha, thanks for the update and all the work you've been doing to improve the test runner :)

File: client/lib/user/support-user-interop.js [hunk 4]
```diff
@@ -36,11 +36,11 @@ const reduxStoreReady = new Promise( ( resolve ) => {
 } );
 export const setReduxStore = _setReduxStore;
 
-/**
- * Check if there's a support user to be activated on boot
- * @return {bool} true if a support user token is waiting to be injected on boot, false otherwise
- */
-export const shouldBoot = () => {
+// Evaluate isSupportUserSession at module startup time, then freeze it
```
Comment (retrofox) [inline client/lib/user/support-user-interop.js]: why did you change the commenting block? using // instead of `/** .. */` my bit :-p

Comment (gwwar) [inline client/lib/user/support-user-interop.js]: It's not jsdoc, so we use `//`. If we don't like this, we should update the guidelines. https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#comments

File: client/lib/user/support-user-interop.js [hunk 5]
```diff
@@ -36,11 +36,11 @@ const reduxStoreReady = new Promise( ( resolve ) => {
...
+// for the remainder of the session. This is needed because the User
+// module clears the store on change; it could return false if called
+// after boot.
+const _isSupportUserSession = () => {
```
Comment (aduth) [inline client/lib/user/support-user-interop.js]: In the course of upgrading to Babel 6 in #1515, I've found this to be invalid syntax. For IIFE with arrow functions, this should be:

```js
const _isSupportUserSession = ( () => { /* ... */ } )();
```

Not:

```js
const _isSupportUserSession = () => { /* ... */ }();
```

Comment (aduth) [inline client/lib/user/support-user-interop.js]: Specifically, see 80347797793fe54ff0a8342b46641c47106b4a06.

File: client/lib/user/support-user-interop.js (hunk 4 above)
Comment (retrofox) [inline client/lib/user/support-user-interop.js]: oh, right. I thought It was the block comment for `_isSupportUserSession` function.

File: client/lib/user/support-user-interop.js (hunk 5 above)
Comment (gwwar) [inline client/lib/user/support-user-interop.js]: Great :eyes:, updated in 3afcc91

**Comment by dllh** [comment]
Gave this a test drive, and it worked as advertised.

**Comment by jordwest** [comment]
Tested and working well with `support-user` feature both enabled and disabled. Thanks for putting this together! :ship:

File: client/state/test/initial-state.js [hunk 6]
```diff
@@ -56,11 +62,60 @@ describe( 'initial-state', () => {
...
+						expect( console.error.called ).to.equal( false );
+					} );
+					it( 'does not build using local forage state', () => {
+						expect( state.postTypes.items[ 2916284 ] ).to.equal( undefined );
+					} );
+					it( 'does not add timestamp to store', () => {
+						expect( state._timestamp ).to.equal( undefined );
+					} );
+					it( 'builds state using server state', () => {
+						expect( state.currentUser.id ).to.equal( 123456789 );
```
Comment (jordwest) [inline client/state/test/initial-state.js]: Is it possible to bypass this as well? This would likely be part of a separate PR. If the server state referred to here comes from data rendered into the page source - we'll want to ignore that. The server has no concept of whether a session is a support user or not - so it'll always render the original user's data. For example with the user bootstrap, we're ignoring the data rendered into the page and doing an XHR with the support token to fetch the user information.

Comment (gwwar) [inline client/state/test/initial-state.js]: :+1: I'll update. Thanks for the catch. cc @seear

Comment (gwwar) [inline client/state/test/initial-state.js]: Added in 6fbdc7c

**Comment by gwwar** [comment]
Going to squash the last few commits
